File: packages/server/src/services/chat-service.ts
import { ChatSchema, ChatMessageSchema, ChatMessage, Chat, ExtendedChatMessage } from 'shared/src/schemas/chat.schemas'
import { ApiError } from 'shared'
import { chatStorage, type ChatsStorage, type ChatMessagesStorage } from '@/utils/storage/chat-storage' // New import
import { ZodError } from 'zod'

export type CreateChatOptions = {
  copyExisting?: boolean
  currentChatId?: string
}
/**
 * Returns an object of functions handling chat logic in a functional style
 * using JSON file storage.
 */
export function createChatService() {
  async function createChat(title: string, options?: CreateChatOptions): Promise<Chat> {
    const chatId = chatStorage.generateId('chat')
    const now = new Date().toISOString()

    const newChatData: Chat = {
      id: chatId,
      title,
      createdAt: now,
      updatedAt: now
    }

    try {
      ChatSchema.parse(newChatData) // Validate before adding to storage
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(`Validation failed for new chat data: ${error.message}`, error.flatten().fieldErrors)
        throw new ApiError(500, `Internal validation error creating chat.`, 'CHAT_VALIDATION_ERROR', error.flatten().fieldErrors)
      }
      throw error // Should not happen if data is constructed correctly
    }

    const allChats = await chatStorage.readChats()

    if (options?.copyExisting && options?.currentChatId) {
      if (!allChats[options.currentChatId]) {
        throw new ApiError(
          404,
          `Referenced chat with ID ${options.currentChatId} not found for copying.`,
          'REFERENCED_CHAT_NOT_FOUND'
        )
      }
    }
    if (allChats[chatId]) {
      // Extremely unlikely with UUIDs but good to check
      throw new ApiError(509, `Chat ID conflict for ${chatId}`, 'CHAT_ID_CONFLICT');
    }

    allChats[chatId] = newChatData
    await chatStorage.writeChats(allChats)
    // Initialize an empty messages file for the new chat
    await chatStorage.writeChatMessages(chatId, {})

    if (options?.copyExisting && options?.currentChatId) {
      const sourceMessages = await chatStorage.readChatMessages(options.currentChatId)
      const messagesToCopy: ChatMessagesStorage = {}
      for (const msg of Object.values(sourceMessages)) {
        const newMessageId = chatStorage.generateId('msg')
        // Preserve original creation timestamp for ordering, assign new ID and new chatId
        const copiedMsg: ChatMessage = {
          ...msg,
          id: newMessageId,
          chatId: chatId,
          // createdAt is already part of msg and should be a valid ISO string
        }
        try {
          ChatMessageSchema.parse(copiedMsg);
          messagesToCopy[newMessageId] = copiedMsg;
        } catch (error) {
          if (error instanceof ZodError) {
            console.error(`Validation failed for copied message ${msg.id} to new chat ${chatId}: ${error.message}`, error.flatten().fieldErrors);
            // Decide whether to skip this message or throw an error for the whole operation
          }
        }
      }
      if (Object.keys(messagesToCopy).length > 0) {
        await chatStorage.writeChatMessages(chatId, messagesToCopy)
      }
    }
    return newChatData
  }

  async function updateChatTimestamp(chatId: string): Promise<void> {
    const allChats = await chatStorage.readChats()
    if (!allChats[chatId]) {
      throw new ApiError(404, `Chat with ID ${chatId} not found for timestamp update.`, 'CHAT_NOT_FOUND')
    }
    allChats[chatId].updatedAt = new Date().toISOString()
    try {
      ChatSchema.parse(allChats[chatId]); // Re-validate before writing
      await chatStorage.writeChats(allChats)
    } catch (error) {
      if (error instanceof ZodError) {
        throw new ApiError(500, `Validation failed updating chat timestamp for ${chatId}.`, 'CHAT_VALIDATION_ERROR', error.flatten().fieldErrors);
      }
      throw error;
    }
  }

  async function saveMessage(message: ExtendedChatMessage): Promise<ExtendedChatMessage> {
    const allChats = await chatStorage.readChats()
    if (!allChats[message.chatId]) {
      throw new ApiError(
        404,
        `Chat with ID ${message.chatId} not found. Cannot save message.`,
        'CHAT_NOT_FOUND_FOR_MESSAGE'
      )
    }

    const messageId = message.id || chatStorage.generateId('msg')
    const now = new Date().toISOString()

    const finalMessageData: ChatMessage = {
      id: messageId,
      chatId: message.chatId,
      role: message.role,
      content: message.content,
      createdAt: message.createdAt || now, // Use provided createdAt if exists (e.g. for imported messages), else new
    }

    try {
      ChatMessageSchema.parse(finalMessageData);
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(`Validation failed for new message data: ${error.message}`, error.flatten().fieldErrors);
        throw new ApiError(500, `Internal validation error saving message.`, 'MESSAGE_VALIDATION_ERROR', error.flatten().fieldErrors);
      }
      throw error;
    }

    const chatMessages = await chatStorage.readChatMessages(message.chatId)
    if (chatMessages[messageId]) {
      // Handle update or throw conflict, for now, let's assume saveMessage can overwrite if ID is provided and exists.
      // Or, more strictly, throw an error if message.id was provided and already exists.
      console.warn(`Message with ID ${messageId} already exists in chat ${message.chatId}. Overwriting.`);
    }
    chatMessages[messageId] = finalMessageData
    await chatStorage.writeChatMessages(message.chatId, chatMessages)
    await updateChatTimestamp(message.chatId)

    return { ...finalMessageData, tempId: message.tempId }
  }

  async function updateMessageContent(chatId: string, messageId: string, content: string): Promise<void> {
    const allChats = await chatStorage.readChats()
    if (!allChats[chatId]) {
      // This check could be redundant if readChatMessages implies chat existence
      throw new ApiError(404, `Chat with ID ${chatId} not found.`, 'CHAT_NOT_FOUND')
    }

    const chatMessages = await chatStorage.readChatMessages(chatId)
    if (!chatMessages[messageId]) {
      throw new ApiError(404, `Message with ID ${messageId} not found in chat ${chatId} for update.`, 'MESSAGE_NOT_FOUND')
    }

    chatMessages[messageId].content = content
    // chatMessages[messageId].updatedAt = new Date().toISOString(); // If messages had an updatedAt field

    try {
      ChatMessageSchema.parse(chatMessages[messageId]); // Re-validate before writing
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(`Validation failed updating message content for ${messageId}: ${error.message}`, error.flatten().fieldErrors);
        throw new ApiError(500, `Internal validation error updating message.`, 'MESSAGE_VALIDATION_ERROR', error.flatten().fieldErrors);
      }
      throw error;
    }

    await chatStorage.writeChatMessages(chatId, chatMessages)
    await updateChatTimestamp(chatId)
  }

  async function getAllChats(): Promise<Chat[]> {
    const allChatsData = await chatStorage.readChats()
    const chatList = Object.values(allChatsData)
    chatList.sort((a, b) => b.updatedAt.localeCompare(a.updatedAt)) // Sort by most recently updated
    return chatList
  }

  async function getChatMessages(chatId: string): Promise<ChatMessage[]> {
    const allChats = await chatStorage.readChats()
    if (!allChats[chatId]) {
      throw new ApiError(404, `Chat with ID ${chatId} not found.`, 'CHAT_NOT_FOUND')
    }

    const chatMessagesData = await chatStorage.readChatMessages(chatId)
    const messageList = Object.values(chatMessagesData)
    messageList.sort((a, b) => a.createdAt.localeCompare(b.createdAt)) // Sort by creation time
    return messageList
  }

  async function updateChat(chatId: string, title: string): Promise<Chat> {
    const allChats = await chatStorage.readChats()
    if (!allChats[chatId]) {
      throw new ApiError(404, `Chat with ID ${chatId} not found for update.`, 'CHAT_NOT_FOUND')
    }

    allChats[chatId].title = title
    allChats[chatId].updatedAt = new Date().toISOString()

    try {
      ChatSchema.parse(allChats[chatId]);
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(`Validation failed updating chat ${chatId}: ${error.message}`, error.flatten().fieldErrors);
        throw new ApiError(500, `Internal validation error updating chat.`, 'CHAT_VALIDATION_ERROR', error.flatten().fieldErrors);
      }
      throw error;
    }
    await chatStorage.writeChats(allChats)
    return allChats[chatId]
  }

  async function deleteChat(chatId: string): Promise<void> {
    const allChats = await chatStorage.readChats()
    if (!allChats[chatId]) {
      throw new ApiError(404, `Chat with ID ${chatId} not found for deletion.`, 'CHAT_NOT_FOUND')
    }

    delete allChats[chatId]
    await chatStorage.writeChats(allChats)
    await chatStorage.deleteChatData(chatId) // Delete the chat's message directory
  }

  // MODIFIED: Added chatId parameter
  async function deleteMessage(chatId: string, messageId: string): Promise<void> {
    const allChats = await chatStorage.readChats()
    if (!allChats[chatId]) {
      throw new ApiError(404, `Chat with ID ${chatId} not found when attempting to delete message ${messageId}.`, 'CHAT_NOT_FOUND')
    }
    const chatMessages = await chatStorage.readChatMessages(chatId)
    if (!chatMessages[messageId]) {
      throw new ApiError(404, `Message with ID ${messageId} not found in chat ${chatId} for deletion.`, 'MESSAGE_NOT_FOUND')
    }

    delete chatMessages[messageId]
    await chatStorage.writeChatMessages(chatId, chatMessages)
    await updateChatTimestamp(chatId)
  }

  async function forkChat(sourceChatId: string, excludedMessageIds: string[] = []): Promise<Chat> {
    const allChats = await chatStorage.readChats()
    const sourceChat = allChats[sourceChatId]
    if (!sourceChat) {
      throw new ApiError(404, `Source chat with ID ${sourceChatId} not found for forking.`, 'SOURCE_CHAT_NOT_FOUND')
    }

    const newTitle = `Fork of ${sourceChat.title} (${new Date().toLocaleTimeString()})`
    const newChatId = chatStorage.generateId('chat')
    const now = new Date().toISOString()
    const newChatData: Chat = {
      id: newChatId,
      title: newTitle,
      createdAt: now,
      updatedAt: now
    }
    ChatSchema.parse(newChatData); // Validate

    allChats[newChatId] = newChatData
    await chatStorage.writeChats(allChats)

    const sourceMessagesAll = await chatStorage.readChatMessages(sourceChatId)
    const messagesToCopyRecord: ChatMessagesStorage = {}

    for (const msg of Object.values(sourceMessagesAll)) {
      if (!excludedMessageIds.includes(msg.id)) {
        const newMessageId = chatStorage.generateId('msg') // New ID for the message in the new chat
        const copiedMsgData: ChatMessage = {
          ...msg,
          id: newMessageId,
          chatId: newChatId, // Point to the new chat
          // createdAt is preserved from original message
        }
        ChatMessageSchema.parse(copiedMsgData); // Validate
        messagesToCopyRecord[newMessageId] = copiedMsgData
      }
    }

    await chatStorage.writeChatMessages(newChatId, messagesToCopyRecord)
    return newChatData
  }

  async function forkChatFromMessage(
    sourceChatId: string,
    messageId: string, // The message ID to fork *up to and including*
    excludedMessageIds: string[] = []
  ): Promise<Chat> {
    const allChats = await chatStorage.readChats()
    const sourceChat = allChats[sourceChatId]
    if (!sourceChat) {
      throw new ApiError(404, `Source chat with ID ${sourceChatId} not found.`, 'SOURCE_CHAT_NOT_FOUND')
    }

    const sourceMessagesAll = await chatStorage.readChatMessages(sourceChatId)
    const startMessage = sourceMessagesAll[messageId]
    if (!startMessage) {
      throw new ApiError(404, `Starting message with ID ${messageId} not found in chat ${sourceChatId}.`, 'MESSAGE_NOT_FOUND')
    }
    // No need for: if (startMessage.chat_id !== sourceChatId) as we fetched from the correct file.

    const newTitle = `Fork from ${sourceChat.title} at message (${messageId.substring(0, 8)})`
    const newChatId = chatStorage.generateId('chat')
    const now = new Date().toISOString()
    const newChatData: Chat = { id: newChatId, title: newTitle, createdAt: now, updatedAt: now }
    ChatSchema.parse(newChatData); // Validate

    allChats[newChatId] = newChatData
    await chatStorage.writeChats(allChats)

    // Get messages up to and including the startMessage
    const sourceMessagesArray = Object.values(sourceMessagesAll).sort((a, b) => a.createdAt.localeCompare(b.createdAt))
    const indexOfStart = sourceMessagesArray.findIndex((m) => m.id === messageId)

    if (indexOfStart === -1) {
      // Should not happen if startMessage was found above
      throw new ApiError(500, `Internal error: Could not re-find the starting message ${messageId} in chat sequence.`, 'MESSAGE_SEQUENCE_ERROR')
    }

    const messagesToConsider = sourceMessagesArray.slice(0, indexOfStart + 1)
    const messagesToCopyRecord: ChatMessagesStorage = {}

    for (const msg of messagesToConsider) {
      if (!excludedMessageIds.includes(msg.id)) {
        const newMessageId = chatStorage.generateId('msg')
        const copiedMsgData: ChatMessage = { ...msg, id: newMessageId, chatId: newChatId }
        ChatMessageSchema.parse(copiedMsgData); // Validate
        messagesToCopyRecord[newMessageId] = copiedMsgData
      }
    }

    await chatStorage.writeChatMessages(newChatId, messagesToCopyRecord)
    return newChatData
  }

  return {
    createChat,
    updateChatTimestamp,
    saveMessage,
    updateMessageContent, // Signature changed
    getAllChats,
    getChatMessages,
    updateChat,
    deleteChat,
    deleteMessage, // Signature changed
    forkChat,
    forkChatFromMessage
  }
}

export const chatService = createChatService()

File: packages/python_backend/app/services/chat_service.py
# app/services/chat_service.py
from datetime import datetime, timezone
import time
from typing import List, Optional
from app.schemas.chat_schemas import (
    Chat, ChatMessage, ExtendedChatMessage, CreateChatBody, MessageRoleEnum, ModelOptions
)
from app.utils.storage.chat_storage import chat_storage
from app.error_handling.api_error import ApiError
from app.core.config import LOW_MODEL_CONFIG
from pydantic import ValidationError

class ChatService:
    async def _update_chat_timestamp(self, chat_id: int) -> None:
        all_chats = await chat_storage.read_chats()
        chat_entry = all_chats.get(chat_id)
        if not chat_entry:
            raise ApiError(status=404, message=f"Chat with ID {chat_id} not found for timestamp update.", code='CHAT_NOT_FOUND')
        now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
        chat_entry = chat_entry.model_copy(update={"updated": now_ms})
        all_chats[chat_id] = chat_entry
        await chat_storage.write_chats(all_chats)

    async def create_chat(self, title: str, copy_existing: Optional[bool] = None, current_chat_id: Optional[int] = None, model_config: Optional[ModelOptions] = None) -> Chat:
        chat_id = chat_storage.generate_id()
        now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
        chat_model_config = model_config or ModelOptions(**LOW_MODEL_CONFIG)
        new_chat_data = Chat(id=chat_id, title=title, created=now_ms, updated=now_ms, chat_model_config=chat_model_config)
        all_chats = await chat_storage.read_chats()
        if copy_existing and current_chat_id:
            if current_chat_id not in all_chats:
                raise ApiError(status=404, message=f"Referenced chat {current_chat_id} not found.", code='REFERENCED_CHAT_NOT_FOUND')
        if chat_id in all_chats:
            raise ApiError(status=409, message=f"Chat ID conflict for {chat_id}", code='CHAT_ID_CONFLICT')

        all_chats[chat_id] = new_chat_data
        await chat_storage.write_chats(all_chats)
        await chat_storage.write_chat_messages(chat_id, {})

        if copy_existing and current_chat_id:
            source_messages = await chat_storage.read_chat_messages(current_chat_id)
            messages_to_copy = {}
            for msg_id, msg_data in source_messages.items():
                new_msg_id = chat_storage.generate_id()
                copied_msg = msg_data.model_copy(update={"id": new_msg_id, "chatId": chat_id})
                messages_to_copy[new_msg_id] = copied_msg
            if messages_to_copy:
                await chat_storage.write_chat_messages(chat_id, messages_to_copy)
        return new_chat_data

    async def save_message(self, message: ExtendedChatMessage) -> ExtendedChatMessage:
        all_chats = await chat_storage.read_chats()
        chat_id = int(message.chat_id) if hasattr(message, 'chat_id') else int(message.chatId)
        if chat_id not in all_chats:
            raise ApiError(status=404, message=f"Chat {chat_id} not found.", code='CHAT_NOT_FOUND_FOR_MESSAGE')

        message_id = message.id or chat_storage.generate_id()
        now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
        valid_role = MessageRoleEnum(message.role) if isinstance(message.role, str) else message.role
        final_message_data = ChatMessage(id=message_id, chatId=chat_id, role=valid_role, content=message.content, created=message.created or now_ms)
        chat_messages = await chat_storage.read_chat_messages(chat_id)
        chat_messages[message_id] = final_message_data
        await chat_storage.write_chat_messages(chat_id, chat_messages)
        await self._update_chat_timestamp(chat_id)
        return ExtendedChatMessage(**final_message_data.model_dump(), temp_id=getattr(message, 'temp_id', None))

    async def get_all_chats(self) -> List[Chat]:
        chats_storage = await chat_storage.read_chats()
        chat_list = list(chats_storage.values())
        chat_list.sort(key=lambda c: c.updated, reverse=True)
        return chat_list

    async def get_chat_messages(self, chat_id: int) -> List[ChatMessage]:
        all_chats = await chat_storage.read_chats()
        if chat_id not in all_chats:
            raise ApiError(status=404, message=f"Chat with ID {chat_id} not found.", code='CHAT_NOT_FOUND')
        messages = await chat_storage.read_chat_messages(chat_id)
        message_list = list(messages.values())
        message_list.sort(key=lambda m: m.created)
        return message_list

    async def update_chat(self, chat_id: int, title: str) -> Chat:
        all_chats = await chat_storage.read_chats()
        if chat_id not in all_chats:
            raise ApiError(status=404, message=f"Chat {chat_id} not found for update.", code='CHAT_NOT_FOUND')
        now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
        updated_chat = all_chats[chat_id].model_copy(update={"title": title, "updated": now_ms})
        all_chats[chat_id] = updated_chat
        await chat_storage.write_chats(all_chats)
        return updated_chat

    async def delete_chat(self, chat_id: int) -> None:
        all_chats = await chat_storage.read_chats()
        if chat_id not in all_chats:
            raise ApiError(status=404, message=f"Chat {chat_id} not found for deletion.", code='CHAT_NOT_FOUND')
        del all_chats[chat_id]
        await chat_storage.write_chats(all_chats)
        await chat_storage.delete_chat_data(chat_id)

    async def delete_message(self, chat_id: int, message_id: int) -> None:
        all_chats = await chat_storage.read_chats()
        if chat_id not in all_chats:
            raise ApiError(status=404, message=f"Chat {chat_id} not found.", code='CHAT_NOT_FOUND')
        messages = await chat_storage.read_chat_messages(chat_id)
        if message_id not in messages:
            raise ApiError(status=404, message=f"Message {message_id} not found in chat {chat_id}.", code='MESSAGE_NOT_FOUND')
        del messages[message_id]
        await chat_storage.write_chat_messages(chat_id, messages)
        await self._update_chat_timestamp(chat_id)

    async def fork_chat(self, source_chat_id: int, excluded_message_ids: List[int] = None) -> Chat:
        excluded_message_ids = excluded_message_ids or []
        all_chats = await chat_storage.read_chats()
        if source_chat_id not in all_chats:
            raise ApiError(status=404, message=f"Source chat {source_chat_id} not found.", code='SOURCE_CHAT_NOT_FOUND')
        source_chat = all_chats[source_chat_id]
        fork_title = f"Fork of {source_chat.title}"
        new_chat = await self.create_chat(fork_title)
        source_messages = await chat_storage.read_chat_messages(source_chat_id)
        messages_to_copy = {}
        for msg_id, msg_data in source_messages.items():
            if msg_id not in excluded_message_ids:
                new_msg_id = chat_storage.generate_id()
                copied_msg = msg_data.model_copy(update={"id": new_msg_id, "chatId": new_chat.id})
                messages_to_copy[new_msg_id] = copied_msg
        if messages_to_copy:
            await chat_storage.write_chat_messages(new_chat.id, messages_to_copy)
        return new_chat

    async def fork_chat_from_message(self, source_chat_id: int, message_id: int, excluded_message_ids: List[int] = None) -> Chat:
        excluded_message_ids = excluded_message_ids or []
        all_chats = await chat_storage.read_chats()
        if source_chat_id not in all_chats:
            raise ApiError(status=404, message=f"Source chat {source_chat_id} not found.", code='SOURCE_CHAT_NOT_FOUND')
        source_chat = all_chats[source_chat_id]
        source_messages = await chat_storage.read_chat_messages(source_chat_id)
        if message_id not in source_messages:
            raise ApiError(status=404, message=f"Message {message_id} not found in chat {source_chat_id}.", code='MESSAGE_NOT_FOUND')
        fork_title = f"Fork of {source_chat.title} from message"
        new_chat = await self.create_chat(fork_title)
        messages_to_copy = {}
        target_message = source_messages[message_id]
        target_timestamp = target_message.created
        for msg_id, msg_data in source_messages.items():
            if (msg_data.created <= target_timestamp and msg_id not in excluded_message_ids):
                new_msg_id = chat_storage.generate_id()
                copied_msg = msg_data.model_copy(update={"id": new_msg_id, "chatId": new_chat.id})
                messages_to_copy[new_msg_id] = copied_msg
        if messages_to_copy:
            await chat_storage.write_chat_messages(new_chat.id, messages_to_copy)
        return new_chat